### PR TITLE
Set instruction_type on UnitGroup and Script

### PR DESCRIPTION
**Pull request by dmcavoy**
Setting the instruction type field for UnitGroups and Scripts. Setting everything to teacher_led unless its called out in [this doc](https://docs.google.com/document/d/1Yx0_MoQTup51fasuNJFvOD3VoRGAZT8_lxDnVffy4Po/edit). A Script only needs to have this field set if it is not in a UnitGroup. If it is in a UnitGroup the UnitGroup instruction_type will be used instead.

## Links

- [Part of Part 1 of Eng Plan](https://docs.google.com/document/d/1V61xONU0-mleMEEdHWNxhvZtYIPRCB31mT_hbUXusCQ/edit#)
- [Product Spec](https://docs.google.com/document/d/1qZpc90daxQiiqpcQUiUiGV2tQoaYEQUzOA72UZzBlVE/edit#)
- https://codedotorg.atlassian.net/browse/PLAT-1346

## Testing story

- Ran script locally

### CODE REVIEW

**Review by davidsbailey — COMMENTED** [review]
Sorry Dani but I don't think this should be run as a migration. it only needs to be run in the levelbuilder environment. then the serialization/seed process will propagate the changes to other environments. if we do this migration, things could get weird during the next deploy to each environment -- first this migration gets run, then the seed step runs, undoing the effects of the migration. the changes won't stick until they appear in the .script_json and .course files.

the simplest thing would probably be to turn this into a oneoff script which you will run from the levelbuilder environment. here is an example you could use as a starting point for that kind of script: https://github.com/code-dot-org/code-dot-org/blob/00e5899210bc66f374ff4864dce1ffe2cab2cdd5/bin/oneoff/find_unseeded_level_concept_difficulty.rb then I would suggest:
1. raise unless levelbuilder_mode 
2.  consider adding to your code to call `write_script_json` or `write_serialization` so that you don't need to go and do that manually after. 
3.  you should be able to test it out pretty well locally (though not perfectly since the stable json sort PR still isn't in yet).

**Comment by dmcavoy** [comment]
@davidsbailey Thanks for the suggestion. I will update it to a oneoff script. 

In the past for things like published_state and other fields I've added we have used a migration to fill them in. Can you help me understand why this is different? Has something changed about the system where now we should plan to use oneoffs in the future? (Looking at instructor and participant audiences which I will be adding soon).

File: bin/oneoff/set_instruction_type.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# for others courses
+
+require_relative '../../dashboard/config/environment'
+
+def set_instruction_type
+  raise unless Rails.application.config.levelbuilder_mode
+
+  Script.all.each do |script|
+    # scripts in unit_groups get their instruction type from their unit group
+    next unless script.unit_group
```
Comment (bethanyaconnor) [inline bin/oneoff/set_instruction_type.rb]: Should this be `next if script.unit_group`?

File: bin/oneoff/set_instruction_type.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+                                'teacher_led'
+                              end
+    script.write_script_json
+  end
+
+  UnitGroup.all.each do |course|
+    # default is teacher_led so only need to update unit groups we want to be self_paced
+    if ['self-paced-pl-csp-2021', 'self-paced-pl-csd-2021'].include?(course.name)
+      course.instruction_type = 'self_paced'
+      course.write_serialization
```
Comment (bethanyaconnor) [inline bin/oneoff/set_instruction_type.rb]: should course also be saved here?

Comment (bethanyaconnor) [reply bin/oneoff/set_instruction_type.rb]: and I guess script above

File: bin/oneoff/set_instruction_type.rb (hunk 1 above)
Comment (dmcavoy) [reply bin/oneoff/set_instruction_type.rb]: Yup! Thank you!

File: bin/oneoff/set_instruction_type.rb (hunk 2 above)
Comment (dmcavoy) [reply bin/oneoff/set_instruction_type.rb]: Added!

**Review by bethanyaconnor — APPROVED** [review]

**Comment by davidsbailey** [comment]
> @davidsbailey Thanks for the suggestion. I will update it to a oneoff script.
> 
> In the past for things like published_state and other fields I've added we have used a migration to fill them in. Can you help me understand why this is different? Has something changed about the system where now we should plan to use oneoffs in the future? (Looking at instructor and participant audiences which I will be adding soon).

I think I just didn't notice that this was not the right pattern until after the published state PRs had shipped. I can't remember exactly but I think there may have been some problems where the migration stopped working on people's development machines because future code changes made the code in the migration no longer correct. when that happened I realized that we probably shouldn't have done it with a migration. sorry for not commenting on it at that time.

in general, changes to levelbuilder content (.course, .script_json and .level files) should get propagated via the serialized format, either by generating changes locally or on levelbuilder.

File: bin/oneoff/set_instruction_type.rb [hunk 3]
```diff
@@ -21,7 +21,7 @@ def set_instruction_type
                               else
                                 'teacher_led'
                               end
-    script.save!
+    script.save(validate: false)
```
Comment (davidsbailey) [inline bin/oneoff/set_instruction_type.rb]: sorry to butt in -- if you want your validations to run on the other changes you are making to the script object, you could also use `skip_name_format_validation`, like this: https://github.com/code-dot-org/code-dot-org/blob/ff2ea1757f704a3526a3c8af0e7c303f8a9aeedc/dashboard/app/models/script.rb#L1318

Comment (davidsbailey) [reply bin/oneoff/set_instruction_type.rb]: I believe this attribute affects whether we validate the script name, but then the parameter itself doesn't get persisted to the database.

Comment (dmcavoy) [reply bin/oneoff/set_instruction_type.rb]: oh nice! i will update

**Pull request merged**